File: runtime/src/main/java/com/siimkinks/sqlitemagic/ExprS.java
package com.siimkinks.sqlitemagic;

import com.siimkinks.sqlitemagic.Select.Select1;
import com.siimkinks.sqlitemagic.internal.SimpleArrayMap;

import java.util.ArrayList;
import java.util.LinkedList;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

final class ExprS extends Expr {
  @NonNull
  private final SelectBuilder<Select1> selectBuilder;
  @Nullable
  private ArrayList<String> parentObservedTables;

  ExprS(@NonNull Column<?, ?, ?, ?, ?> column,
        @NonNull String op,
        @NonNull SelectSqlNode.SelectNode<?, Select1, ?> selectNode) {
    super(column, op);
    this.selectBuilder = selectNode.selectBuilder;
  }

  @Override
  void addArgs(@NonNull ArrayList<String> args) {
    super.addArgs(args);
    args.addAll(selectBuilder.args);
  }

  @Override
  void addObservedTables(@NonNull ArrayList<String> tables) {
    super.addObservedTables(tables);
    // defer observed tables adding until sql building, where we might add missing joins,
    // so there will be more tables to add/observe
    this.parentObservedTables = tables;
  }

  @Override
  void appendToSql(@NonNull StringBuilder sb) {
    super.appendToSql(sb);
    sb.append('(');
    selectBuilder.appendCompiledQuery(sb, parentObservedTables);
    sb.append(')');
  }

  @Override
  void appendToSql(@NonNull StringBuilder sb, @NonNull SimpleArrayMap<String, LinkedList<String>> systemRenamedTables) {
    super.appendToSql(sb, systemRenamedTables);
    sb.append('(');
    selectBuilder.appendCompiledQuery(sb, parentObservedTables);
    sb.append(')');
  }
}

File: runtime/src/main/kotlin/com/siimkinks/sqlitemagic/ExprS.kt
package com.siimkinks.sqlitemagic

import com.siimkinks.sqlitemagic.Select.Select1
import com.siimkinks.sqlitemagic.SelectSqlNode.SelectNode
import com.siimkinks.sqlitemagic.internal.SimpleArrayMap
import java.util.LinkedList

internal class ExprS(
  column: Column<*, *, *, *, *>,
  op: String,
  selectNode: SelectNode<*, Select1, *>
) : Expr(
  column = column,
  op = op
) {
  private val selectBuilder = selectNode.selectBuilder
  private var parentObservedTables: ArrayList<String>? = null

  override fun addArgs(args: ArrayList<String?>) {
    super.addArgs(args)
    args.addAll(selectBuilder.args)
  }

  override fun addObservedTables(tables: ArrayList<String>) {
    super.addObservedTables(tables)
    // defer observed tables adding until sql building, where we might add missing joins,
    // so there will be more tables to add/observe
    parentObservedTables = tables
  }

  override fun appendToSql(sb: StringBuilder) {
    super.appendToSql(sb)
    sb.append('(')
    selectBuilder.appendCompiledQuery(sb, parentObservedTables)
    sb.append(')')
  }

  override fun appendToSql(
    sb: StringBuilder,
    systemRenamedTables: SimpleArrayMap<String, LinkedList<String>>
  ) {
    super.appendToSql(sb, systemRenamedTables)
    sb.append('(')
    selectBuilder.appendCompiledQuery(sb, parentObservedTables)
    sb.append(')')
  }
}
